storagenode/retain/retain.go

```diff
@@ -88,105 +89,209 @@ type Service struct {
+		s.group.Go(func() error {
+			// Grab lock to check things.
+			s.cond.L.Lock()
+			defer s.cond.L.Unlock()
+
+			for {
+				// If we have closed, exit.
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-s.closed:
+					return nil
+				default:
+				}
 
-			if queueLength == 0 {
-				s.emptyTrigger <- struct{}{}
-			}
+				// Grab next item from queue.
+				request, ok := s.next()
+				if !ok {
+					// Nothing in queue, go to sleep and wait for
+					// things shutting down or next item.
+					s.cond.Wait()
+					continue
+				}
+
+				// Temporarily Unlock so others can work on the queue while
+				// we're working on the retain request.
+				s.cond.L.Unlock()
+
+				// Signal to anyone waiting that the queue state changed.
+				s.cond.Broadcast()
 
-			// remove item from semaphore and free up process for another retain job
-			<-s.semaphore
-		}(ctx, req)
+				// Run retaining process.
+				err := s.retainPieces(ctx, request)
+				if err != nil {
+					s.log.Error("retain pieces failed", zap.Error(err))
+				}
+
+				// Mark the request as finished. Relock to maintain that
+				// at the top of the for loop the lock is held.
+				s.cond.L.Lock()
+				s.finish(request)
+				s.cond.Broadcast()
+			}
+		})
 	}
+
+	// Unlock while we wait for the workers to exit.
+	s.cond.L.Unlock()
+	err := s.group.Wait()
+	s.cond.L.Lock()
+
+	// Clear the queue after Wait has exited. We're sure no more entries
+	// can be added after we acquire the mutex because wait spawned a
+	// worker that ensures the closed channel is closed before it exits.
+	s.queued = nil
+	s.cond.Broadcast()
+
+	return err
 }
 
-// Wait blocks until the context is canceled or until the queue is empty.
-func (s *Service) Wait(ctx context.Context) {
-	s.mu.Lock()
-	queueLength := len(s.queued)
-	s.mu.Unlock()
-	if queueLength == 0 {
-		return
+// next returns next item from queue, requires mutex to be held
+func (s *Service) next() (Request, bool) {
+	for id, request := range s.queued {
+		// Check whether a worker is retaining this satellite,
+		// if, yes, then try to get something else from the queue.
+		if _, ok := s.working[request.SatelliteID]; ok {
+			continue
+		}
+		delete(s.queued, id)
+		// Mark this satellite as being worked on.
+		s.working[request.SatelliteID] = struct{}{}
+		return request, true
 	}
-	select {
-	case <-s.emptyTrigger:
-	case <-ctx.Done():
+	return Request{}, false
+}
+
+// finish marks the request as finished, requires mutex to be held
+func (s *Service) finish(request Request) {
+	delete(s.working, request.SatelliteID)
+}
+
+// Close causes any pending Run to exit and waits for any retain requests to
+// clean up.
+func (s *Service) Close() error {
+	s.cond.L.Lock()
+	s.closedOnce.Do(func() { close(s.closed) })
+	s.cond.L.Unlock()
+
+	s.cond.Broadcast()
+	_ = s.group.Wait()
+	return nil
+}
+
+// Wait blocks until the queue and working is empty. When Run exits, it empties the queue.
+func (s *Service) Wait() {
```

Comment:
We should still call this `TestWaitUntilEmpty` since it is only used for tests @zeebo

Reply:
Ah, I was unaware of that. I'll change it back.

Reply:
The body is also different:

```
// TestWaitUntilEmpty blocks until the context is canceled or until the queue is empty.
func (s *Service) TestWaitUntilEmpty(ctx context.Context, pollInterval time.Duration) {
	ticker := time.NewTicker(pollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			s.cond.L.Lock()
			count := len(s.queued)
			s.cond.L.Unlock()


			if count == 0 {
				return
			}
		case <-ctx.Done():
			return
		}
	}
}
```

Reply:
And in the tests we passed in `30 * time.Millisecond` as `pollInterval`

Reply:
Also `TestWaitUntilEmpty` still needs to handle `ctx`.

Reply:
Why? When the context is canceled, Run exits, which empties the queue, which unblocks Wait.